ui-app/client/src/components/PageEditor/components/FormEditor.tsx

```diff
@@ -0,0 +1,289 @@
+import { duplicate, isNullValue } from '@fincity/kirun-js';
+import React, { useCallback, useEffect, useState } from 'react';
+import Portal from '../../Portal';
+import { IconHelper } from '../../util/IconHelper';
+import CommonCheckbox from '../../../commonComponents/CommonCheckbox';
+import { LocationHistory, PageDefinition } from '../../../types/common';
+import { PageStoreExtractor, getDataFromPath, setData } from '../../../context/StoreContext';
+import { getHref } from '../../util/getHref';
+import axios from 'axios';
+import { shortUUID } from '../../../util/shortUUID';
+import { LOCAL_STORE_PREFIX } from '../../../constants';
+import { compDefinitionGenerator, generateFormPreview } from '../functions/CompDefgenerator';
+import { generateFormEvents } from '../functions/EventDefGenerator';
+
+interface FormName {
+	appCode: string;
+	clientCode: string;
+	id: number;
+	name: string;
+}
+
+interface FormEditorProps {
+	formStorageUrl: string;
+	defPath: string | undefined;
+	pageExtractor: PageStoreExtractor;
+	locationHistory: Array<LocationHistory>;
+	clickedComponent: string;
+	setClickedComponent: (pref: any) => void;
+}
+export default function FormEditor({
+	formStorageUrl,
+	defPath,
+	pageExtractor,
+	locationHistory,
+	clickedComponent,
+	setClickedComponent,
+}: FormEditorProps) {
+	const [currentForm, setCurrentForm] = useState<any>();
+	const [eachFormDef, setEachFormDef] = useState<any>({});
+	const [selectedFormComponent, setSelectedFormComponent] = useState<Array<string>>([]);
+	const [addForm, setAddForm] = useState<boolean>(false);
+	const [formName, setFormName] = useState<string>('');
+	const [formDefs, setFormDefs] = useState<FormName[]>();
+	const [showFormFields, setShowFormFields] = useState<boolean>(false);
+	const [addSubmitButton, setAddSubmitButton] = useState<boolean>(false);
+	const [addClearButton, setAddClearButton] = useState<boolean>(false);
+
+	const submitButtonKey = shortUUID();
+
+	const clearButtonKey = shortUUID();
+
+	const headers: any = {
+		Authorization: getDataFromPath(`${LOCAL_STORE_PREFIX}.AuthToken`, []),
+	};
+	if (globalThis.isDebugMode) headers['x-debug'] = shortUUID();
+
+	const callForFormDef = useCallback(() => {
```

Comment:
Don't use call back. Create a separate function outside the component to make the axios calls. Same thing goes for the one in the bottom.

Reply:
Also have to do the pagination